### DISCUSSION TITLE
Seamless debugging between python and Rust?

### DISCUSSION THREAD

**Discussion by topherbuckley** [PyO3/pyo3#2997] [General]
Hello, I just started looking into a project that appears to use PyO3 and I've never used Rust before so wondering how people are going about debugging such projects. 

I read through the [debugging section](https://pyo3.rs/main/debugging.html) of the docs, but it seems the gdb usecase there is running a rust binary as the entry point rather than python. And the python example is just a single function test and I can't access the python source from within rust-gdb it seems. 

I'm no expert, but I am somewhat familiar with the debugging tools available in gdb, PyCharm and VSCode, but not sure if there is a way to inline debug (or even print things) seamlessly between the python and rust code. 

Taking the README example of `sum_as_string`, I can start the Pycharm debugger and step into the generated stub of `string_sum.py`, but its just a stub. I'm looking for some kind of debugging tools that would allow me to either inline debug and go from python to rust and back again, or some way to print things (variable values, types, etc.) from rust while running from python. 

The `string_sum` example is fairly straightforward to understand, but I'm trying to get a grasp on a MUCH larger project. I don't necessarily need to know all the innards of the project, as I just want to create a PR to fix one bug affecting me as an end user. Trying to find a good way to track down the source of such a bug by creating breakpoints near the identified methods controlling the user-input and following them thereafter. 

Without being intimately familiar with a given project, what is a good way to step through the code of a PyO3 project like this to get a feel for how it works?

**Comment by adamreichold** [comment]
This will not improve the debugging experience w.r.t. the Python code, but you should be able to attach GDB to the CPython process executing your top-level Python code and it still place break points within any native extensions the interpreter loaded. (This should also work with `rust-gdb` which is just a wrapper loading some printing helpers.)

But you can certainly also apply "printf-style" debugging to the Rust code itself which should be visible on standard output like calling Python's `print` if nothing fancy is done to that. There is also a macro called [`dbg!`](https://doc.rust-lang.org/stable/std/macro.dbg.html) to make this painless in the standard library.

**Reply by topherbuckley** [reply]
@adamreichold thanks for the tips! I tried the following for the `sum_as_string` example in the README:

1. `rust-gdb python` (where python is the active venv version)
2. `break lib.rs:12`
>No source file named lib.rs.
>Make breakpoint pending on future shared library load? (y or [n]) y
>Breakpoint 1 (lib.rs:12) pending.
3. `run main.py` (contents of which shown below)
```python
import string_sum

sum = string_sum.sum_as_string(5,20)
print(sum)
```
4. gdb breaks at lib.rs:12 as expected

Now this is all fine and dandy, but I know where to set the breakpoint in this simple example. What about in projects where you don't know which rust file contains the definition for some function used in python? I.e. I know my entry point in the python side, but don't know the linking Rust file. What is a good way to find that? Again, hoping for a way to find this without being intimately familiar with the whole code base. 

The printf-style and `dbg!` debugging would also require me to already know where to start looking in the Rust code right? I'm not seeing a way around the catch-22 of already having to know what I'm trying to find out. :D Any further tips appreciated!

**Reply by adamreichold** [reply]
If you at least know the module name, you can break on its initialization function, e.g. `break string_sum::string_sum::init` in your example.

If you don't know that, you should be able to break on the generic `_PyImport_LoadDynamicModuleWithSpec` (if you have debug information for CPython) and `finish` this so that the Rust module will be loaded after that.

**Reply by adamreichold** [reply]
As a sort of meta comment: If you do have a procedure that works for you when native extensions written in C/C++ are involved, you should able to apply this without modifications to native extensions written in Rust. The resulting machine code and debug information are the same. (The rust-gdb wrapper will only auto-load the Python scripts in e.g. `$RUSTUP_HOME/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/etc" which should be able to use in any GDB-based environment and there are similar scripts available for LLDB in that same directory.)

**Reply by kngwyu** [reply]
Using debug build of CPython (e.g., you can get it by pyenv) would help but I don't think there's any magic... debugging is just as hard as other programs. Also as adamreichold points out having C/C++ extension that compiled without debugging info can make the situation hard.

**Reply by topherbuckley** [reply]
### init entry
@adamreichold rust-gdb is not breaking anywhere when I used `break string_sum::string_sum::init`, 

Note I am just using the `Cargo.toml` and `src/lib.rs` from the README as copied below. Does there need to be an explicit `init` function definition in the rust or python code or something like that? :

`Cargo.toml`
```rust
[package]
name = "string_sum"
version = "0.1.0"
edition = "2018"

[lib]
# The name of the native library. This is the name which will be used in Python to import the
# library (i.e. `import string_sum`). If you change this, you must also change the name of the
# `#[pymodule]` in `src/lib.rs`.
name = "string_sum"
# "cdylib" is necessary to produce a shared library for Python to import from.
#
# Downstream Rust code (including code in `bin/`, `examples/`, and `tests/`) will not be able
# to `use string_sum;` unless the "rlib" or "lib" crate type is also included, e.g.:
# crate-type = ["cdylib", "rlib"]
crate-type = ["cdylib"]

[dependencies]
pyo3 = { version = "0.18.1", features = ["extension-module"] }
```

`src/lib.rs`
```rust
use pyo3::prelude::*;

/// Formats the sum of two numbers as string.
#[pyfunction]
fn sum_as_string(a: usize, b: usize) -> PyResult<String> {
    Ok((a + b).to_string())
}

/// A Python module implemented in Rust. The name of this function must match
/// the `lib.name` setting in the `Cargo.toml`, else Python will not be able to
/// import the module.
#[pymodule]
fn string_sum(_py: Python<'_>, m: &PyModule) -> PyResult<()> {
    m.add_function(wrap_pyfunction!(sum_as_string, m)?)?;
    Ok(())
}
```

So far I've tried setting breakpoints at the following:

```
(gdb) info breakpoints 
Num     Type           Disp Enb Address            What
1       breakpoint     keep y   <PENDING>          break string_sum::string_sum::init
2       breakpoint     keep y   <PENDING>          break string_sum::string_sum
3       breakpoint     keep y   <PENDING>          break string_sum::init
4       breakpoint     keep y   <PENDING>          break string_sum
5       breakpoint     keep y   0x00007ffff73e5527 in string_sum::sum_as_string 
                                                   at src/lib.rs:6

```
Where the `<PENDING>` ones obviously did not find a matching source. This is at least somewhat helpful, as I can potentially give gdb the rust function name. If you can provide any further tips on how to use the init entry point, it would be appreciated. 

### _PyImport_LoadDynamicModuleWithSpec entry

I'm able to break using `break _PyImport_LoadDynamicModuleWithSpec`, but not quite sure how to figure out the entry point to lib.rs from there. For example:

```python
rust-gdb python
(gdb) break _PyImport_LoadDynamicModuleWithSpec
Breakpoint 1 at 0x149ac0: file ./Python/importdl.c, line 94.
(gdb) run main.py 
Starting program: /home/christopher/Downloads/string_sum/.env/bin/python main.py
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".

Breakpoint 1, _PyImport_LoadDynamicModuleWithSpec (spec=spec@entry=0x7ffff7485280, fp=fp@entry=0x0) at ./Python/importdl.c:94
94	{
(gdb) l
89	    return NULL;
90	}
91	
92	PyObject *
93	_PyImport_LoadDynamicModuleWithSpec(PyObject *spec, FILE *fp)
94	{
95	#ifndef MS_WINDOWS
96	    PyObject *pathbytes = NULL;
97	#endif
98	    PyObject *name_unicode = NULL, *name = NULL, *path = NULL, *m = NULL;
(gdb) inf args
spec = 0x7ffff7485280
fp = 0x0
(gdb) finish
Run till exit from #0  _PyImport_LoadDynamicModuleWithSpec (spec=spec@entry=0x7ffff7485280, fp=fp@entry=0x0) at ./Python/importdl.c:94
_imp_create_dynamic_impl (module=<optimized out>, file=<optimized out>, spec=0x7ffff7485280) at Python/import.c:2304
2304	    Py_DECREF(name);
Value returned is $1 = (PyObject *) 0x7ffff748a220
(gdb) l
2299	    else
2300	        fp = NULL;
2301	
2302	    mod = _PyImport_LoadDynamicModuleWithSpec(spec, fp);
2303	
2304	    Py_DECREF(name);
2305	    Py_DECREF(path);
2306	    if (fp)
2307	        fclose(fp);
2308	    return mod;
```

Maybe I have to follow through a few more calls with `finish` but I'm not really sure what I am supposed to be looking for. Should the `file` arg within `_PyImport_LoadDynamicModuleWithSpec` be the `lib.rs` file I'm interested in finding?

**Reply by adamreichold** [reply]
> Where the <PENDING> ones obviously did not find a matching source.

This is to be expected as you need to actually dynamically load the extension before those breakpoints can be resolved. When you start the Python interpreter, before the `import string_sum` statement is executed, the extension is not loaded yet.

> Maybe I have to follow through a few more calls with finish but I'm not really sure what I am supposed to be looking for. 

The point here is at the end of `_PyImport_LoadDynamicModuleWithSpec`, the native extension is loaded and you should be able to inspect it sources from within GDB. This would give you a first starting point if you do not know anything about the source structure of the extension.

**Reply by topherbuckley** [reply]
Maybe I've misunderstood, but I though that running through `run main.py` within gdb would have loaded the extension. Otherwise I don't really understand how the breakpoint 5 @ `string_sum::sum_as_string` would have been resolved. I should have clarified that I ran `info breakpoints` after running `run main.py` and it finished execution. After which, the `lib.rs` could be resolved within gdb. 

>you should be able to inspect it sources from within GDB.

You mean find it directly, or find it within the list of `info sources`? Based on what you wrote last time, I was under the impression that `finish` would result in me entering the `lib.rs` file, but I think I've misunderstood that.

**Reply by adamreichold** [reply]
> Maybe I've misunderstood, but I though that running through run main.py within gdb would have loaded the extension. Otherwise I don't really understand how the breakpoint 5 @ string_sum::sum_as_string would have been resolved. I should have clarified that I ran info breakpoints after running run main.py and it finished execution. After which, the lib.rs could be resolved within gdb.

Yes, if you ran that after running the script, the breakpoint to the `init` function should have been resolved. Not sure why GDB does find the function but does find `sum_as_string`, it works here:

```console
string_sum> rust-gdb python3
GNU gdb (GDB; openSUSE Tumbleweed) 12.1
[..]
Reading symbols from python3...
(No debugging symbols found in python3)
Missing separate debuginfos, use: zypper install python310-base-debuginfo-3.10.9-3.1.x86_64
(gdb) break string_sum::string_sum::init
Function "string_sum::string_sum::init" not defined.
Make breakpoint pending on future shared library load? (y or [n]) y
Breakpoint 1 (string_sum::string_sum::init) pending.
(gdb) r
Starting program: /usr/bin/python3 
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib64/libthread_db.so.1".
Python 3.10.9 (main, Dec 08 2022, 14:49:06) [GCC] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import string_sum

Breakpoint 1, string_sum::string_sum::init () at src/lib.rs:12
12      #[pymodule]
(gdb) bt
#0  string_sum::string_sum::init () at src/lib.rs:12
#1  0x00007ffff7dcb8e6 in ?? () from /lib64/libpython3.10.so.1.0
#2  0x00007ffff7d2cb1b in ?? () from /lib64/libpython3.10.so.1.0
#3  0x00007ffff7d24269 in _PyEval_EvalFrameDefault () from /lib64/libpython3.10.so.1.0
#4  0x00007ffff7d208a3 in ?? () from /lib64/libpython3.10.so.1.0
#5  0x00007ffff7d25ea0 in _PyEval_EvalFrameDefault () from /lib64/libpython3.10.so.1.0
#6  0x00007ffff7d208a3 in ?? () from /lib64/libpython3.10.so.1.0
#7  0x00007ffff7d21d47 in _PyEval_EvalFrameDefault () from /lib64/libpython3.10.so.1.0
#8  0x00007ffff7d208a3 in ?? () from /lib64/libpython3.10.so.1.0
#9  0x00007ffff7d219ca in _PyEval_EvalFrameDefault () from /lib64/libpython3.10.so.1.0
#10 0x00007ffff7d208a3 in ?? () from /lib64/libpython3.10.so.1.0
#11 0x00007ffff7d219ca in _PyEval_EvalFrameDefault () from /lib64/libpython3.10.so.1.0
#12 0x00007ffff7d208a3 in ?? () from /lib64/libpython3.10.so.1.0
#13 0x00007ffff7d219ca in _PyEval_EvalFrameDefault () from /lib64/libpython3.10.so.1.0
#14 0x00007ffff7d208a3 in ?? () from /lib64/libpython3.10.so.1.0
#15 0x00007ffff7d2c14a in ?? () from /lib64/libpython3.10.so.1.0
#16 0x00007ffff7d3786e in _PyObject_CallMethodIdObjArgs () from /lib64/libpython3.10.so.1.0
#17 0x00007ffff7d370ed in PyImport_ImportModuleLevelObject () from /lib64/libpython3.10.so.1.0
#18 0x00007ffff7d24ee7 in _PyEval_EvalFrameDefault () from /lib64/libpython3.10.so.1.0
#19 0x00007ffff7d208a3 in ?? () from /lib64/libpython3.10.so.1.0
#20 0x00007ffff7d974c0 in PyEval_EvalCode () from /lib64/libpython3.10.so.1.0
#21 0x00007ffff7da89a3 in ?? () from /lib64/libpython3.10.so.1.0
#22 0x00007ffff7da423a in ?? () from /lib64/libpython3.10.so.1.0
#23 0x00007ffff7d0491b in ?? () from /lib64/libpython3.10.so.1.0
#24 0x00007ffff7d04a90 in _PyRun_InteractiveLoopObject () from /lib64/libpython3.10.so.1.0
#25 0x00007ffff7c95a75 in _PyRun_AnyFileObject () from /lib64/libpython3.10.so.1.0
#26 0x00007ffff7d04c01 in PyRun_AnyFileExFlags () from /lib64/libpython3.10.so.1.0
#27 0x00007ffff7c8bf57 in ?? () from /lib64/libpython3.10.so.1.0
#28 0x00007ffff7d893b7 in Py_BytesMain () from /lib64/libpython3.10.so.1.0
#29 0x00007ffff7a2caf0 in __libc_start_call_main () from /lib64/libc.so.6
#30 0x00007ffff7a2cbb9 in __libc_start_main_impl () from /lib64/libc.so.6
#31 0x0000555555555075 in _start ()
(gdb) 
```

(Using a debug build of the crate.)

> You mean find it directly, or find it within the list of info sources? Based on what you wrote last time, I was under the impression that finish would result in me entering the lib.rs file, but I think I've misunderstood that.

Yes, I just meant that you can intercept the Python interpreter loading a PyO3 extension on which you basically have no prior knowledge. But indeed, you could also single step your way into the extensions initialization function this way.

**Reply by topherbuckley** [reply]
Huh, I also am no longer having an issue with it. I can't reproduce what I was seeing before  ¯\_(ツ)_/¯. Thanks for confirming though. Can you generalize the syntax you used for that though? i.e. where are the three terms in `string_sum::string_sum::init` defined? Would this be `<Cargo.toml-package>::<Cargo.tom-lib>::init`? Why is it `init` vs `__init__` or something else? Happy to read relevant documentation, so if this is something specific to rust/cargo that I need to read up on, please just let me know that.

**Reply by kngwyu** [reply]
`init` is generated by our macro (https://github.com/PyO3/pyo3/blob/v0.18.1/pyo3-macros-backend/src/module.rs#L86), so it's just an ad-hoc naming by us. `<crate-name>::<crate-name>::init` should work. Here `crate-name` means `lib.name` in Cargo.toml.
(The second one can be different for non-top-level modules, but you might not need to care)

**Comment by kngwyu** [comment]
Do you mean that you want to set the breakpoint in Python code and jump into Rust code from that break point? I don't know whether there's any Python debugger that works like that.

**Reply by topherbuckley** [reply]
Ideally yes, but the workaround to use rust-gdb to run the python code (but limited to inspecting rust code) is also helpful.

**Comment by squidfunk** [comment]
I'm probably late to the party, but given the very valuable information in this thread, I was successful to set up proper Python/Rust debugging in Visual Studio Code. Maybe it's of use for somebody, because I found it hard to get accurate information on this.

Here's a sample project:
https://github.com/squidfunk/pyo3-debug-vscode

**Comment by wangzhen0518** [comment]
The plugin [Python C++ Debugger](https://marketplace.visualstudio.com/items?itemName=benjamin-simmonds.pythoncpp-debug) may satisfy your requirements. I found this plugin while debugging python and cpp implemented with [pybind11](https://github.com/pybind/pybind11). Since debugging rust program is similar to debugging cpp programs, I think this plugin may also work when debugging python and rust, and I succeeded.

This is a screenshot of a successful debugging.
![image](https://github.com/user-attachments/assets/7b55f581-38f6-4d22-9eaa-048d0a007730)

Here is my configuration in `launch.json`.
```json
{
    "version": "0.2.0",
    "configurations": [
        {
            "name": "pyo3 debug",
            "type": "pythoncpp",
            "request": "launch",
            "entirePythonConfig": {
                "name": "Python: Current File",
                "type": "python",
                "request": "launch",
                "program": "${file}",
                "python": "${workspaceFolder}/.venv/bin/python",
                "console": "integratedTerminal"
            },
            "entireCppConfig": {
                "name": "(lldb) Attach",
                "type": "lldb",
                "request": "attach",
                "processId": "",
                "program": "${workspaceFolder}/.venv/lib/python3.11/site-packages/rs4py/rs4py.cpython-311-x86_64-linux-gnu.so",
                "sourceLanguages": [
                    "rust"
                ]
            },
        }
    ]
}
```

Hope this can solve your problem.
